src/protocols/secure_channel/CASESession.cpp

```diff
@@ -982,104 +900,79 @@ CHIP_ERROR CASESession::HandleSigmaR3(System::PacketBufferHandle & msg)
-    SuccessOrExit(err);
-
-    err = tlvReader.FindElementWithTag(CASETLVTag::kEncryptedData, suppTlvReader);
-    SuccessOrExit(err);
-    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_ByteString, err = CHIP_ERROR_WRONG_TLV_TYPE);
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-    VerifyOrExit(msg_R3_Encrypted.Alloc(suppTlvReader.GetLength()), err = CHIP_ERROR_NO_MEMORY);
-    msg_r3_encrypted_len = static_cast<uint16_t>(suppTlvReader.GetLength());
-    err                  = suppTlvReader.GetBytes(msg_R3_Encrypted.Get(), msg_r3_encrypted_len);
-    SuccessOrExit(err);
-
-    err = tlvReader.FindElementWithTag(CASETLVTag::kTag, suppTlvReader);
-    SuccessOrExit(err);
-    VerifyOrExit(suppTlvReader.GetType() == TLV::kTLVType_ByteString, err = CHIP_ERROR_WRONG_TLV_TYPE);
-    VerifyOrExit(kTAGSize == suppTlvReader.GetLength(), err = CHIP_ERROR_INVALID_TLV_ELEMENT);
-    err = suppTlvReader.GetBytes(tag, sizeof(tag));
-    SuccessOrExit(err);
+    // Fetch encrypted data
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, CHIP_ERROR_INVALID_TLV_TAG);
+    VerifyOrExit(msg_R3_Encrypted.Alloc(tlvReader.GetLength()), err = CHIP_ERROR_NO_MEMORY);
+    msg_r3_encrypted_len = static_cast<uint16_t>(tlvReader.GetLength());
+    SuccessOrExit(err = tlvReader.GetBytes(msg_R3_Encrypted.Get(), msg_r3_encrypted_len));
 
     // Step 1
     saltlen = kIPKSize + kSHA256_Hash_Length;
 
     VerifyOrExit(msg_salt.Alloc(saltlen), err = CHIP_ERROR_NO_MEMORY);
 
-    err = ComputeIPK(mConnectionState.GetPeerKeyID(), mRemoteIPK, sizeof(mRemoteIPK));
-    SuccessOrExit(err);
-
     {
         MutableByteSpan saltSpan(msg_salt.Get(), saltlen);
-        err = ConstructSaltSigmaR3(mRemoteIPK, sizeof(mRemoteIPK), saltSpan);
-        SuccessOrExit(err);
+        SuccessOrExit(err = ConstructSaltSigmaR3(ByteSpan(mIPK, sizeof(mIPK)), saltSpan));
     }
 
-    err = mHKDF.HKDF_SHA256(mSharedSecret, mSharedSecret.Length(), msg_salt.Get(), saltlen, kKDFSR3Info, kKDFInfoLength, sr3k,
-                            kAEADKeySize);
-    SuccessOrExit(err);
+    SuccessOrExit(err = mHKDF.HKDF_SHA256(mSharedSecret, mSharedSecret.Length(), msg_salt.Get(), saltlen, kKDFSR3Info,
+                                          kKDFInfoLength, sr3k, kAEADKeySize));
 
-    err = mCommissioningHash.AddData(buf, bufLen);
-    SuccessOrExit(err);
+    SuccessOrExit(err = mCommissioningHash.AddData(buf, bufLen));
 
-    // Step 2
-    err = AES_CCM_decrypt(msg_R3_Encrypted.Get(), msg_r3_encrypted_len, nullptr, 0, tag, kTAGSize, sr3k, kAEADKeySize, kIVSR3,
-                          kIVLength, msg_R3_Encrypted.Get());
-    SuccessOrExit(err);
+    // Step 2 - Decrypt data blob
+    SuccessOrExit(err = AES_CCM_decrypt(msg_R3_Encrypted.Get(), msg_r3_encrypted_len - kTAGSize, nullptr, 0,
```

Comment:
Need to precheck that `msg_r3_encrypted_len > kTAGSize`, same for R2

Reply:
Also, the multiple indices could be resolved with adding a helper variable:

```
size_t msg_r3_encrypted_len_with_tag = 0;
size_t msg_r3_encrypted_len = 0;
......
msg_r3_encrypted_len_with_tag = tlvReader.GetLength();
VerifyOrExit(msg_r3_encrypted_len_with_tag > kTAGSize, CHIP_ERROR_INVALID_TLV_ELEMENT);
msg_r3_encrypted_len = msg_r3_encrypted_len_with_tag - kTAGSize;
...

SuccessOrExit(err = AES_CCM_decrypt(msg_R3_Encrypted.Get(), msg_r3_encrypted_len, nullptr, 0,
    msg_R3_Encrypted.Get() + msg_r3_encrypted_len, kTAGSize, sr3k, kAEADKeySize,
    kTBEData3_Nonce, kTBEDataNonceLength, msg_R3_Encrypted.Get()));
....
```

Reply:
Updated.

Reply:
Updated.